Adding an Ipyleaflet wrapper for WFS support

f-PLT opened the pull request:
## Overview
As dissussed in this issue https://github.com/Ouranosinc/pavics-sdi/issues/159

Small wrapper module to simplify and automate the use of a WFS with ipyleaflet

Changes:

New sub-module in birdy, with example notebooks.

## To verify
Unsure sure about where to put the new requirement. For now, I added Ipyleaflet to the requirements.txt and environment.yml

I'm not sure how or where (or even if) I have to change something for the new module to be included in a "Birdy" future release.

f-PLT commented:
Working on the Codacy issues

huard commented on birdy/ipyleafletwfs/README.md:
```diff
@@ -0,0 +1,18 @@
+
+This module is to be used inside a jupyter notebook, either with a standard server or through vscode/pycharm.
+
+There are a notebook examples which show how to use this module and what can be done with it.
```
```suggestion
There are notebook examples which show how to use this module and what can be done with it.
```

huard commented on birdy/ipyleafletwfs/README.md:
```diff
@@ -0,0 +1,18 @@
+
+**Warning**
+WFS requests and GeoJSON layers can take a long time to process and render. Trying to load a lake layers for a nation wide extent will take a long time
+and probably crash. The denser and complex the layer, the more zoomed in the map extent should be.
```
```suggestion
and probably crash. The more dense and complex the layer, the more zoomed-in the map extent should be.
```

huard commented on birdy/ipyleafletwfs/base.py:
```diff
@@ -0,0 +1,394 @@
+          ex: public:canada_forest_layer
+
+        source_map: Map instance
+          The map instance on which the layer is to be added
```
```suggestion
          The map instance on which the layer is to be added.
```

huard commented on birdy/ipyleafletwfs/base.py:
```diff
@@ -0,0 +1,394 @@
+          The map instance on which the layer is to be added
+
+        layer_style: dictionnary
+          ipyleaflet GeoJSON style format. See ipyleaflet documentation for more information
+
+          ex: { 'color': 'white', 'opacity': 1, 'dashArray': '9', 'fillOpacity': 0.1, 'weight': 1 }
```
```suggestion
          ipyleaflet GeoJSON style format, for example `{ 'color': 'white', 'opacity': 1, 'dashArray': '9', 'fillOpacity': 0.1, 'weight': 1 }`. See ipyleaflet documentation for more information. 

          
```

huard commented on birdy/ipyleafletwfs/base.py:
```diff
@@ -0,0 +1,394 @@
+
+        property: string
+          The property key to be used by the widget. Use the property_list() function
+          to get a list of the available properties
```
```suggestion
          to get a list of the available properties.
```

Zeitsperre replied:
```suggestion
          to get a list of the available properties.
```

huard commented on birdy/ipyleafletwfs/base.py:
```diff
@@ -0,0 +1,394 @@
+
+        Will create a GeoJSON layer, filtered by the extent of the source_map parameter.
+        If no source map is given, it will not filter by extent, which can cause problems
+        with large layers
```
```suggestion
        with large layers.
```

Zeitsperre commented on birdy/ipyleafletwfs/README.md:
```diff
@@ -0,0 +1,18 @@
+Using the on-map 'Refresh WFS layer' button will make a new request for the current extent.
+
+**Warning**
+WFS requests and GeoJSON layers can take a long time to process and render. Trying to load a lake layers for a nation wide extent will take a long time
```
```suggestion
WFS requests and GeoJSON layers are costly operations to process and render. Trying to load lake layers at the nationwide extent may take a long time
```

Zeitsperre commented on birdy/ipyleafletwfs/base.py:
```diff
@@ -0,0 +1,394 @@
+        self._property_widgets.clear()
+
+    def create_feature_property_widget(self, widget_name, property=None, widget_position='bottomright'):
+        """Create a visualization widget for a specific feature property
```
```suggestion
        """Create a visualization widget for a specific feature property.
```

Zeitsperre commented on birdy/ipyleafletwfs/base.py:
```diff
@@ -0,0 +1,394 @@
+    def create_feature_property_widget(self, widget_name, property=None, widget_position='bottomright'):
+        """Create a visualization widget for a specific feature property
+
+        Will create a widget for the layer and source map used to create
```
?
```suggestion
        Will create a widget for the layer and source map.
```

f-PLT commented:
Perfect. I'll get to work on the current revision and the Codacy check in the mean time

Zeitsperre commented on environment.yml:
```diff
@@ -27,4 +27,5 @@ dependencies:
   - sphinx
   - nbsphinx
   - nbconvert
+  # ipyleaflet
```
```suggestion
```

tlogan2000 commented on birdy/ipyleafletwfs/examples/ipyleafletwfs_guide.ipynb:
```diff
@@ -0,0 +1,342 @@
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "wfs_connection.remove_layer()"
```
@huard  asked me to take a look at this as well ... Have started here runnning on a local env and everything ran up to the `remove_layer()` call. Traceback :

```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-15-4d3049f72d4a> in <module>
----> 1 wfs_connection.remove_layer()

~/github/github_birdy/birdy/ipyleafletwfs/base.py in remove_layer(self)
    224         if self._layer:
    225             # Remove maps elements
--> 226             self.clear_property_widgets()
    227             self._source_map.remove_control(self._refresh_widget)
    228             self._source_map.remove_layer(self._layer)

~/github/github_birdy/birdy/ipyleafletwfs/base.py in clear_property_widgets(self)
    341           The map instance from which the widgets are to be removed.
    342         """
--> 343         for widget in self._property_widgets:
    344             self._source_map.remove_control(self._property_widgets[widget]['widget'])
    345         self._property_widgets = None

TypeError: 'NoneType' object is not iterable
```

f-PLT replied:
Thanks for that catch. Forgot to change this after another change. 

Adding a conditional check to clear_property_widgets() to prevent this

f-PLT commented:
Thank you for the review. If you think of any more use cases or small quality of life changes, please don't hesitate!

One thing to keep in mind is that leaflet.js (the core of ipyleaflet) is a quick and simple approach to web maps. Simple to setup and launch, harder to do complex things and customize. (The Openlayers javascript library is a lot more suited to custom maps, but there's no equivalent to ipyleaflet for it)

Highlighting the clicked feature should be pretty easy to do as there are already examples for ipyleaflet that do this.

Selecting multiple polygons might be a problem, not sure it is possible with ipyleaflet as is. The "selection" in itself, is a click event that returns the feature's information under it (and the click events are a bit limited as is). It's not so much a selection in the classic GIS sense. It would probably be possible, but require a lot of work, especially in conjuction with the information widgets.

There's probably an ipywidget more suited to tables and longer lists of property items (outside of ipyleaflet, but still compatible).  The information rapidly takes over the map area with the HTML object used in this version, making it impractical to have a lot of elements in it. 

In a normal ipyleaflet approach, the user would build a custom widget for their needs. That said, it will be possible to improve on this, but the main difficulty will be to determine what we want in the information widget and finding the right balance between automation and user control/choice, with what is possible with ipyleaflet.

I think we also have to keep in mind the scope we want for this. I see this could rapidly reach a point where we'll be trying to build, in essence, an interactive web map portal inside a jupyter notebook by automating on top of ipyleaflet, when building a web client with openlayers would be more appropriate, easier and cleaner, or (small GIS analyst bias here) simply teach and/or put together a tutorial on how to use QGIS as a WFS client.

That said, I definitely think there is a lot we can still do with this approach.

tlogan2000 commented:
Ok thanks.

> In a normal ipyleaflet approach, the user would build a custom widget for their needs.

In regards to my table comment,  I'm not sure if I was super clear when describing my idea. In my albeit limited exploration of ipywidgets I found it was relatively very easily combine widgets from different libraries (e.g. the map via ipyleaflet, plus a table created from library {x?}) using the basic layout tools. https://ipywidgets.readthedocs.io/en/latest/examples/Widget%20Styling.html

Code would look something like this 

```
from ipywidgets import  VBox
final_widget = VBox([demo_map, table])
display(final_widget)
```

cehbrecht commented:
`ipyleaflet` adds the notebook and `ipywidget` dependencies. We already got notebook dependencies with `pytest-notebook` ... Previously the notebook dependencies were kept separately in `notebooks/environment.yml`. The core dependencies (command line usage) didn't use the notebook.

Should we skip the separation of the notebook dependencies? Or clean it up (with another PR) and separate them again?

huard commented:
I suggest we keep the notebook related dependencies separate, as I consider these optional to birdy.

f-PLT commented:
In that case, should I also take the ipyleafletwfs module out of the birdy module/folder? ipyleafletwfs  also needs the ipyleaflet dependency, it's not just in the notebooks examples.

huard commented:
Take a look at `birdy/dependencies.py` to see how the `ipywidget` dependency is handled. I suggest we do something similar for this, i.e. put additional wfs related dependencies in `requirements_extra.txt`.

f-PLT commented:
I cleaned up the dependencies and added an entry in dependences.py, but there seems to be an issue with it being an unused import in the codacy analysis.

I tried to fix it with # noqa: F401, as I've seen throughout Birdy, but it did not work..

Was there an issue with ipywidgets and IPython too in dependencies.py?

I could also remove my entry in dependencies.py and manage it in ipyleafletwfs.

huard commented:
I can live with codacy failures.

f-PLT commented:
Alright, I'll be merging at the end of the day, in case any body else has other modification requests

f-PLT commented:
I forgot, I don't have write access, so I can't merge it myself...
